### PR TITLE
Use upstream version of multihash instead of a fork

### PR DESCRIPTION
**Pull request by vmx**
The changes from the libp2p fork have been backported to upstream, hence
upstream can now be used instead.

### CODE REVIEW

**Comment by vmx** [comment]
Is there anything I can help with running it against a live system or is this something the Rust libp2p team will take care of?

**Comment by tomaka** [comment]
> Is there anything I can help with running it against a live system or is this something the Rust libp2p team will take care of?

If you're motivated, my idea was to run [Polkadot](https://github.com/paritytech/polkadot) with a local override to this PR, and see if nothing suspicious happens.
It might be difficult to define "suspicious" if you've never used Polkadot before, but no errors printed on stdout and 25 peers is what I'd look for.

The most difficult part is waiting for the compilation to finish.

**Comment by vmx** [comment]
Can I opt-out again :) I thought there might be something automated, you'd need to ping people for. I've never used Polkadot, so I guess that isn't a good task for me as I really can't tell if there's anything suspicious.

**Comment by tomaka** [comment]
Looks like it's working.

For note-keeping, here's the small patch to apply to Substrate:

```diff
diff --git a/client/authority-discovery/src/lib.rs b/client/authority-discovery/src/lib.rs
index 6260ac9a8..ff88765cb 100644
--- a/client/authority-discovery/src/lib.rs
+++ b/client/authority-discovery/src/lib.rs
@@ -499,9 +499,8 @@ where
 }
 
 fn hash_authority_id(id: &[u8]) -> Result<libp2p::kad::record::Key> {
-	libp2p::multihash::encode(libp2p::multihash::Hash::SHA2256, id)
-		.map(|k| libp2p::kad::record::Key::new(&k))
-		.map_err(Error::HashingAuthorityId)
+	let digest = libp2p::multihash::Sha2_256::digest(id);
+	Ok(libp2p::kad::record::Key::new(&digest))
 }
 
 fn interval_at(start: Instant, duration: Duration) -> Interval {
```

File: core/src/peer_id.rs [hunk 1]
```diff
@@ -69,17 +70,16 @@ impl PeerId {
...
+        let (hash_algorithm, canonical_algorithm): (_, Option<Code>) = /*if key_enc.len() <= MAX_INLINE_KEY_LENGTH {
             (multihash::Hash::Identity, Some(multihash::Hash::SHA2256))
         } else {*/
-            (multihash::Hash::SHA2256, None);
+            (Code::Sha2_256, None);
         //};
 
         let canonical = canonical_algorithm.map(|alg|
-            multihash::encode(alg, &key_enc).expect("SHA2256 is always supported"));
+            alg.hasher().expect("canonical hasher exists").digest(&key_enc));
```
[the remark this answers is not in the document]
Comment (vmx) [reply core/src/peer_id.rs]: If the canonical hasher is always sha2-256, should perhaps then the variable also be renamed?

File: core/src/peer_id.rs [hunk 2]
```diff
@@ -69,17 +70,16 @@ impl PeerId {
...
 
-        let multihash = multihash::encode(hash_algorithm, &key_enc)
-            .expect("identity and sha2-256 are always supported by known public key types");
+        let multihash = hash_algorithm.hasher().expect("hasher exists").digest(&key_enc);
```
Comment (vmx) [inline core/src/peer_id.rs]: @twittner should I also change this message back to "identity and sha2-256 are always supported by known public key types"?

Comment (twittner) [reply core/src/peer_id.rs]: I would go with "identity and sha2-256 are always supported".

File: core/src/peer_id.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (twittner) [reply core/src/peer_id.rs]: "canonical" as a variable name sounds good to me so I would leave it as is. After all SHA2-256 is our canonical hash algorithm.

**Comment by vmx** [comment]
I have just rebased an changed the error messages in https://github.com/libp2p/rust-libp2p/commit/c9769306aa7114c429e55aeb38af6ed99d630814#diff-d21ea1befb9128247a52e731c8fd3256R79-R83 as suggested by the code review.

**Comment by vmx** [comment]
The integration test also fails for me locally on master.

**Comment by vmx** [comment]
I've no idea what the problem is, the integration test works locally.

**Comment by tomaka** [comment]
The integration test failure is obviously not caused by this PR, and we also had it in the other PR.
Let's merge and open an issue.

**Pull request merged**